[doc] Changing docs for repository permissions (#48039)
sourcegraph/sourcegraph: doc/admin/permissions/syncing.md

# Permission syncing

> Permission syncing is a polling mechanism, that periodically checks via an API call to the code host to determine, what are the permissions of a specific entity.

Reviewer: ```suggestion
Permission syncing is a polling mechanism, that periodically checks via an API call to the code host to determine which permissions a specific entity has.
```

> From the point of view of which entity we ask for, we do have 2 ways to sync permissions and both are on by default, resulting in double polling:

Reviewer: ```suggestion
We have two ways to sync permissions. Both are on by default, resulting in double polling:
```

### Mapping code host identifiers

To identify which Sourcegraph users or repositories the permissions relate to, we need to map the users identifiers from code host with userIDs on Sourcegraph and similarly for repository identifiers from code host to repoIDs on Sourcegraph side. 

> WARNING: For this process to work correctly, the user needs to have an external account from code host mapped to the user account on sourcegraph, otherwise the code host identifier cannot be matched and repository permissions cannot be enforced. 

This is the main reason to require users to connect to their code host. This can be done on the Account security settings page: `/users/$USER/settings/security`.

### Entities that do not exist on Sourcegraph

There might be cases, when the identifiers on the code host do not match any user or repository on Sourcegraph side. 
In that case, Sourcegraph still stores the permission information in the internal database as **pending permissions**
In case that an entity with the same code host identifier is added later, Sourcegraph applies these pending permissions.

> > NOTE: Sourcegraph only keeps pending permissions for repositories that were added to Sourcegraph. We will never store information about repositories that are not shared with us.

Reviewer 2: This feels like another place I would remove the Note and just have the text in paragraph form.

If everything works correctly in such a case, [on-demand sync request](#on-demand-sync) is not strictly needed. 
But the permissions of the entity might have changed in the time since the last sync of these pending permissions, 
so it is safer to make the request anyway to keep the permissions as fresh as possible

## Checking permissions sync state

The state of an user or repository's permissions can be checked in the UI by:

- For users: navigating to `/users/$USER/settings/permissions`
- For repositories: navigating to `/$CODEHOST/$REPO/-/settings/permissions`

The GraphQL API can also be used:

```graphql
query {
  user(username: "user") {
    permissionsInfo {
      syncedAt
      updatedAt
    }
  }
}
```

Or for the sake of repository:
```graphql
query {
  repository(name: "repository") {
    permissionsInfo {
      syncedAt
      updatedAt
    }
  }
}
```

In the GraphQL API, `syncedAt` indicates the last complete sync and `updatedAt` indicates the last incremental sync. If `syncedAt` is more recent than `updatedAt`, the user or repository is in a state of complete sync - [learn more](#complete-sync-vs-incremental-sync).

## Sync duration

> When syncing permissions from code hosts with large numbers of users and repositories, it can take a lot of time to complete syncing all permissions from a code host for every user and every repository. Typically due to internal rate limits imposed by Sourcegraph and external rate limits imposed by the code host, it can be hours or days. 

Reviewer 2: The "hours or days" part has me worried. It is honest in the most extreme cases but wondering if it is being too honest. 

What about something like "it can take hours for customers with large number of users and/or repositories. Please contact the support team for further assistance"

For initial setup of instance, when the initial sync for all repositories and users is running, users will gradually see more and more search results from repositories they have access to.

### Lag-time

Let's call the time difference between applying the permissions change on code host and the change taking effect on Sourcegraph as **lag time**. 
For security reasons, we strive to make the lag time as low as possible. 

However given the way the system works, we need to be aware of the worst case:

> IMPORTANT: If there is a change in permissions on the code host, in the worst case the lag time is as long as the time it takes to completely sync all user or repository permissions.

**Example**
There are `5 000` users, `40 000` repositories and the github.com API is paginated on `100` items per page. 
On average, every user has read access to `300` repositories and on average a repository is accessible by `75` users.

User `alice` is removed from repository `horsegraph/not-so-global` at 15:02. How long does it take for this change to take effect on Sourcegraph side?

> Let's say rate limiting is not slowing down permission syncing and default settings for permission syncing are used (scheduler runs every 15 seconds and it schedules 10 users with oldest permissions). That means, we schedule 40 users per minute to be synced. To sync all the users takes $5000 / 40 = 125$ minutes.

Reviewer: Where do the $ signs come from here? 😄

Author: It's supposed to be original Markdown formatting for math formulas. I am trying to figure out right now if our library supports it. If not, I'll rip the dollar signs away.

Default values are shown below:

```json
{
  // Time interval between each iteration of the scheduler
  "permissions.syncScheduleInterval": 15,
  // Number of user permissions to schedule for syncing in single scheduler iteration.
  "permissions.syncOldestUsers": 10,
  // Number of repo permissions to schedule for syncing in single scheduler iteration.
  "permissions.syncOldestRepos": 10,
  // Don't sync a user's permissions if they have synced within the last n seconds.
  "permissions.syncUsersBackoffSeconds": 60,
  // Don't sync a repo's permissions if it has synced within the last n seconds.
  "permissions.syncReposBackoffSeconds": 60,
  // The maximum number of user-centric permissions syncing jobs that can be spawned concurrently.
  // Service restart is required to take effect for changes.
  "permissions.syncUsersMaxConcurrency": 1,
}
```

If the purpose is to fire more requests to the code host, the internal rate limit or the code host rate limit must be also changed accordingly. 

Internal rate limiter settings are described on each code host configuration page, but in general, the `requestsPerHour` field needs to be set to the desired number.

### Recommendations

> #### Count users << Count repositories

Reviewer: ```suggestion
#### Less users than repositories
```

> If there are much less users, than repositories, it is better to rely on user-centric perms sync instead of repo-centric sync. In that case, we recommend:

Reviewer: ```suggestion
If there are a lot less users than repositories, it is better to rely on user-centric perms sync instead of repo-centric sync. In that case, we recommend:
```

```json
  "permissions.syncOldestUsers": 20,
  "permissions.syncOldestRepos": 1,
```

This configuration change will schedule `20` users on each scheduler iteration and just `1` repository. That way, we use the API requests better and user sync will be prefered.
This is just an example, please change the `syncOldestUsers` value to what is desired in your organization. 

**Example**:
There are `10 000` users, `40 000` repositories and the desired time to do a full cycle of permission sync is `2 hours`.
That means we need to sync 5000 users an hour. If we keep the `syncScheduleInterval` to `15s` (the default), we schedule 4-times a minute. `5000/(4 * 60) = 20.8`, so the scheduler needs to schedule 21 users on each iteration to get under the 2 hour mark.

The rate limit for the code host would need to be changed to support the load. In that case the recommendation is to set it to 2x of the amount of [requests expected from permission syncing](#request-count).

> #### Count repositories << Count users

Reviewer: ```suggestion
#### More repositories than users
```

Author: This is actually reversed `More users than repositories`, but fixed so that it aligns with the intent :)
